include/boost/geometry/policies/robustness/segment_ratio.hpp

```diff
@@ -335,12 +342,17 @@ private :
     // Boost.Rational is used if the approximations are close.
     // Reason: performance, Boost.Rational does a GCD by default and also the
     // comparisons contain while-loops.
-    fp_type m_approximation;
+    floating_point_type m_approximation;
 
+    inline bool close_to(thistype const& other) const
+    {
+        return geometry::math::abs(m_approximation - other.m_approximation)
+                < 50;
+    }
 
-    static inline fp_type scale()
+    static inline floating_point_type scale()
     {
-        return 1000000.0;
+        return 1000000;
```

Comment:
On platforms with 16 bit int this would be incorrect. Why did you change it from `double` anyway? I thought that we assume that user-defined numeric types has to be implicitly constructible from `double` (which btw also means that they are implicitly constructible from `int`).

Reply:
> > I changed it because of your suggestion
> 
> Which one? :)

_"Since you moved it we may address its problem now. The problem is that user-defined FP type is going to be degraded to double so this typedef should rather be"_

Reply:
> Do you mean that this is a problem for user-defined floating point type which may not be implicitly constructible from `double`? Which type exactly this fails to compile for? I think `multiprecision::number` should be fine, the ctors are not marked as explicit. They may be disabled/enabled with type traits though. I tested your branch with `1000000.0` and the tests compile with clang-win for me.

I just retested, it is not the Boost multiprecision **float** but the Boost multi precision **int** which gave the problem. I test that in the other branch. But I now realize that it's not enough, also that type should go to floating point. I will fix that later in this PR. And you are right then, this could stay as it was before.

Reply:
So you were talking about FP typedef, not the integral literal.

There are 4 issues at play here:
1. The type `floating_point_type`
2. whether or not it defines explicit ctor
3. whether or not it is constructible from double
4. the type of the literal

2. and 3. may be more complex because of ctors enabled with `enable_if`.

My initial remark was about the last point. I was wrong and the type is not going to be `int` but any integral type able to hold this number. So now I think that any literal you use here should be fine. This doesn't mean that this literal could be used in construction of the `floating_point_type`.

I agree we shouldn't assume that the UD type can be implicitly constructible. This means that in all places ctor should be explicitly called.

> I just retested, it is not the Boost multiprecision float but the Boost multi precision int which gave the problem.

Ok, so this was one of the multiprecision integral types which is not "promoted" to `double` because of the change of `floating_point_type`. For this we probably needs these traits.

Note that if it's not possible to create integral multipresion number from double or float then such numbers will not compile in other parts of the code where we create FP constants.

Reply:
And once again, integer constant is fine here. So this may stay as it is now.